File: src/molsturm/MolecularSystem.cc
//
// Copyright (C) 2017 by the molsturm authors
//
// This file is part of molsturm.
//
// molsturm is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// molsturm is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with molsturm. If not, see <http://www.gnu.org/licenses/>.
//

#include "MolecularSystem.hh"

namespace molsturm {

MolecularSystem::MolecularSystem(gint::Structure structure_, double charge_,
                                 size_t multiplicity_)
      : structure(std::move(structure_)),
        n_alpha(0),
        n_beta(0),
        charge(std::move(charge_)) {
  // Compute number of electrons to distribute to alpha and beta spin:
  const double n_elec = structure.total_charge() - charge;
  const size_t n_elec_count = static_cast<size_t>(n_elec);
  assert_throw(n_elec - n_elec_count < 1e-12, ExcNonIntegerElectronCount(n_elec));

  // Twice the total spin (i.e. the multiplicity reduced by one)
  const size_t spin_twice = [&multiplicity_, &n_elec_count]() {
    if (multiplicity_ == linalgwrap::Constants<size_t>::invalid) {
      // Set to 0 if even number of electrons and 1 if odd
      // equals multiplicities 1 and 2, respectively
      return n_elec_count % 2 == 0 ? size_t(0) : size_t(1);
    } else {
      return multiplicity_ - 1;
    }
  }();

  assert_throw(
        multiplicity_ == linalgwrap::Constants<size_t>::invalid || multiplicity_ > 0,
        ExcInvalidMultiplicity(multiplicity_,
                               "The multiplicity needs to be a positive number."));
  assert_throw(
        spin_twice <= n_elec_count,
        ExcInvalidMultiplicity(
              multiplicity_,
              "The multiplicity should be larger or equal to the electron count (" +
                    std::to_string(n_elec_count) + ") plus 1"));

  const bool even_elec = n_elec_count % 2 == 0;
  const bool even_2spin = spin_twice % 2 == 0;

  // Odd spin_twice values imply a half-integer total spin
  // In other words we need an odd number of electrons in this case
  //
  // Even spin_twice values imply an integer total spin
  // i.e. an even number of electrons
  assert_throw(even_2spin == even_elec,
               ExcInvalidMultiplicity(
                     multiplicity_,
                     "Only a system with an even number of electrons can have a odd "
                     "multiplicity and only a system with an odd number of electrons "
                     "can have an even multiplicity. This system has " +
                           std::to_string(n_elec_count) + " electrons."));

  // Compute number of alpha and beta electrons
  n_alpha = (n_elec_count - spin_twice) / 2 + spin_twice;
  n_beta = n_elec_count - n_alpha;

  // Check that this makes sense
  if (multiplicity_ == linalgwrap::Constants<size_t>::invalid) {
    assert_internal((n_alpha - n_beta) == spin_twice);
  } else {
    assert_internal((n_alpha - n_beta) + 1 == multiplicity_);
  }
}

}  // namespace molsturm

File: src/interface/python/molsturm/MolecularSystem.py
#!/usr/bin/env python3
## vi: tabstop=4 shiftwidth=4 softtabstop=4 expandtab
## ---------------------------------------------------------------------
##
## Copyright (C) 2017 by the molsturm authors
##
## This file is part of molsturm.
##
## molsturm is free software: you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published
## by the Free Software Foundation, either version 3 of the License, or
## (at your option) any later version.
##
## molsturm is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with molsturm. If not, see <http://www.gnu.org/licenses/>.
##
## ---------------------------------------------------------------------

from gint import Structure


class MolecularSystem(Structure):
    """
    Class representing a molecular system, i.e. a structure and a number of electrons.
    """
    def __init__(self, atoms=None, coords=None, electrons=None, multiplicity=None,
                 charge=None, structure=None):
        """
        Initialise a molecular system. By default it is empty (i.e. contains no atoms
        or electrons). In order to construct a non-empty system, specify one or more
        of the optional parameters. Certain combinations are not valid and will
        raise a ValueError or TypeError.

        electrons       The number of electrons in the system.
                        Can be given as a tuple (n_alpha, n_beta) or as a single
                        value. Then we take n_alpha = electrons // 2 and
                        n_beta = electrons - n_alpha
        multiplicity    The multiplicity (2S+1) of the ground state to compute.
                        By default 1 is chosen for even electron systems and 2
                        for odd electron systems.
        charge          Computed from the electrons and the nuclei in the system
                        or set to 0 (neutral atom).
        """

        if structure is None:
            super().__init__(atoms, coords)
        elif not isinstance(structure, Structure):
            raise TypeError("structure needs to be a gint.Structure object.")
        else:
            super().__init__(structure.atom_numbers, structure.coords)

        if electrons is None:
            if charge is None:
                charge = 0
            n_elec_count = super().total_charge - charge

            if multiplicity is None:
                if n_elec_count % 2 == 0:
                    multiplicity = 1
                else:
                    multiplicity = 2
            elif multiplicity <= 0:
                raise ValueError("The multiplicity needs to be a positive number.")
            spin_twice = multiplicity - 1

            if spin_twice % 2 != n_elec_count % 2:
                raise ValueError("Only a system with an even number of electrons can "
                                 "have an odd multiplicity and vice versa. This system "
                                 "has " + str(n_elec_count) + " electrons, but a "
                                 "multiplicity of " + str(multiplicity) + ".")

            self.n_alpha = (n_elec_count - spin_twice) // 2 + spin_twice
            self.n_beta = n_elec_count - self.n_alpha

            assert self.multiplicity == multiplicity
            assert self.n_alpha - self.n_beta == spin_twice
        elif isinstance(electrons, tuple):
            self.n_alpha, self.n_beta = electrons
        else:
            self.n_alpha = electrons // 2
            self.n_beta = electrons - self.n_alpha

        self.__charge = super().total_charge - self.n_alpha - self.n_beta

    @property
    def multiplicity(self):
        """Return the multiplicity of the system"""
        return self.n_alpha - self.n_beta + 1

    @property
    def is_closed_shell(self):
        """Is the system closed shell"""
        return self.n_alpha == self.n_beta

    @property
    def total_charge(self):
        """Return the total resulting charge of the system"""
        return self.__charge

    @property
    def charge(self):
        return self.__charge
    charge.__doc__ = total_charge.__doc__
